maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelProcessor.java

```diff
@@ -63,32 +75,110 @@
+            Path pomFile, InputStream input, Reader reader, Map<String, ?> options) throws IOException {
+        Source source = (Source) options.get(ModelProcessor.SOURCE);
+        if (pomFile == null && source != null) {
+            Path p = Paths.get(source.getLocation());
+            if (Files.exists(p)) {
+                pomFile = p;
+            }
+        }
+        if (pomFile != null) {
+            Path projectDirectory = pomFile.getParent();
+            Path path = pomFile;
+            List<ModelParser> parsers = modelParsers.stream()
+                    .filter(p -> path.equals(p.locatePom(projectDirectory)))
+                    .collect(Collectors.toList());
+            List<ModelParserException> exceptions = new ArrayList<>();
+            for (ModelParser parser : parsers) {
+                try {
+                    Model model = parser.parse(new FileSource(pomFile), options);
+                    return model.withPomFile(pomFile);
+                } catch (ModelParserException e) {
+                    exceptions.add(e);
+                }
+            }
+            try {
+                return readXmlModel(pomFile, null, null, options);
+            } catch (IOException e) {
+                exceptions.forEach(e::addSuppressed);
+                throw e;
+            }
+        } else {
+            return readXmlModel(pomFile, input, reader, options);
+        }
+    }
+
+    private org.apache.maven.api.model.Model readXmlModel(
+            Path pomFile, InputStream input, Reader reader, Map<String, ?> options) throws IOException {
+        if (pomFile != null) {
+            return modelReader.read(pomFile.toFile(), options).getDelegate();
+        } else if (input != null) {
+            return modelReader.read(input, options).getDelegate();
+        } else {
+            return modelReader.read(reader, options).getDelegate();
+        }
     }
 
     @Override
-    public Model read(File input, Map<String, ?> options) throws IOException {
-        return reader.read(input, options);
+    public org.apache.maven.model.Model read(File file, Map<String, ?> options) throws IOException {
+        Objects.requireNonNull(file, "file cannot be null");
+        Path path = file.toPath();
+        org.apache.maven.api.model.Model model = read(path, null, null, options);
+        return new org.apache.maven.model.Model(model);
     }
 
     @Override
-    public Model read(Reader input, Map<String, ?> options) throws IOException {
-        return reader.read(input, options);
+    public org.apache.maven.model.Model read(InputStream input, Map<String, ?> options) throws IOException {
+        Objects.requireNonNull(input, "input cannot be null");
+        try (InputStream in = input) {
+            org.apache.maven.api.model.Model model = read(null, in, null, options);
+            return new org.apache.maven.model.Model(model);
+        } catch (ModelParserException e) {
```

Comment:
I had to read this twice. ModelParserException and ModelParseException are confusingly similarly named since they differ by a single letter.

Reply:
I agree, but that's always the problem to bridge between the v3 and v4 apis.